### PR TITLE
Refactor how plugin dependencies are installed for npm 5

### LINKED ISSUE #2045: Plugin deps fail to install with npm v5
I saw this when submitting #2043, but it took me a while to validate what was going on.  This seems to apply only to npm v5 and only on the invocation of `girder-install web --all-plugins`.  The output from the initial install is [here](https://gist.github.com/jbeezley/d40fc53421cbb05852c3c788533ef85e).  After running, the dependencies (vega in this case) are not installed in `node_modules`.  Here, the failure is in the jobs plugin, but similar behavior appears for any plugin installing extra dependencies.

What is weird is that I have found that you can run `girder-install web --plugins jobs` to fix the error building the jobs plugin.  After that, running with `--all-plugins` will fail on the ldap plugin due to the missing sortablejs.  And again, you can fix it by building that plugin alone.  To get, `--all-plugins` to succeed, you have to first install each plugin individually.

### PR DESCRIPTION
**Pull request by jbeezley**
@girder/developers This is a fairly invasive change, but I attempted to maintain compatibility.  The core of the change is to group all of the `npm install` commands into a single grunt task which is executed before all others.  There is some fairly complex logic that goes into how and where the dependencies are installed, and I think I covered them, but it would be worthwhile for others to test this change on their plugins to be sure.

Fixes #2045

### CODE REVIEW

**Comment by zachmullen** [comment]
Thanks for tackling this so quickly!

File: grunt_tasks/plugin.js [hunk 1]
```diff
@@ -61,6 +61,168 @@ module.exports = function (grunt) {
...
+                })
+            );
+        }
+
+        var child = child_process.spawnSync('npm', args, {
+            stdio: 'inherit'
+        });
+
+        if (child.status !== 0) {
+            throw new Error(`"npm ${args.join(' ')}" failed`);
```
Comment (zachmullen) [inline grunt_tasks/plugin.js]: I'm run into a case where throwing a new Error inside one of these grunt modules does not actually cause a build failure, but I can't remember if it was at config time or task run time. Does this cause the build to fail? If not, `grunt.fail()` will work.

Comment (jbeezley) [inline grunt_tasks/plugin.js]: That's a good point.  I'll check.

File: grunt_tasks/plugin.js [hunk 2]
```diff
@@ -468,6 +537,26 @@ module.exports = function (grunt) {
         configurePluginForBuilding(path.resolve(grunt.config.get('pluginDir'), name), true);
     });
 
+    // This is noop that exists for backwards compatibility in case any plugins rely
+    // on its existence.
+    grunt.registerTask(
```
Comment (zachmullen) [inline grunt_tasks/plugin.js]: Let's add `@deprecated: remove in v3` here. I grep -i for "deprecated" to find things to remove :)

**Comment by zachmullen** [comment]
This works in my testing. It still prunes the tree, so there is some performance downside to rapidly switching between building different sets of plugins, but that's not strictly a problem.

**Comment by jbeezley** [comment]
I suspect that npm will add an option to avoid autoprunning eventually, but for now we just have to deal with it.  Thankfully, installation is quite a bit faster now.

**Comment by zachmullen** [comment]
> Thankfully, installation is quite a bit faster now.

Awesome -- will we get the benefits of this caching in CircleCI already, or do we need to add additional cache dirs?

**Comment by zachmullen** [comment]
Also, should we switch CircleCI to use npm5? If it's easy, I'd vote to do it.

**Comment by jbeezley** [comment]
That seems like a good idea.  It follows how we test with the latest version of pip as well.  The cache is in a different place now, but it is still in `~/.npm`, so circle should still save it.

**Comment by jbeezley** [comment]
@girder/developers Since I haven't heard of any breakages, I'll go ahead and merge.  Let me know if there are any issues.

**Pull request merged**